Add Route choice assignment from path files

Jake-Moss opened the pull request:
This PR allows users to save path files, and then load them back into memory to perform a full assignment at a later date. This includes support for recomputing the path-sized logit with new graph costs and/or new PSL values, using the existing probabilities, conducting a select link analysis, and performing link loading.

This allows using routes from a Pandas DataFrame or a PyArrow hive dataset from disk.

This PR also includes patches for a handful of OOB access and memory leaks. I've also committed my debug and [sanitizer](https://github.com/google/sanitizers/wiki/AddressSanitizer) options I had laying around in my setup.py. ASan required `LD_PRELOAD=libasan.so.8` when running.

#### TODO
- [x] Update documentation
- [x] Tests

Jake-Moss committed:
Fix OOB read in heap when size is 0

Jake-Moss committed:
Avoid passing empty vectors to pyarrow, it just complains about it

Jake-Moss committed:
Docs

Jake-Moss committed:
Grab the correct schema when writing single execution results

Jake-Moss committed:
Fix a handful of leaks in BFS-LE

Jake-Moss committed:
Add RC assignment from path files

Jake-Moss committed:
Add tests for RC path-files

Jake-Moss committed:
Prevent trying to write before results are computed

Jake-Moss committed:
Add small snippet to examples

Jake-Moss committed:
Typo

Jake-Moss commented on aequilibrae/paths/route_choice.py:
```diff
@@ -47,7 +47,7 @@ class RouteChoice:
 
     def __init__(self, graph: Graph, project=None):
-        self.parameters = self.default_parameters.copy()
+        self.parameters = {}
```
`self.parameters = self.default_parameters.copy()` doesn't make sense as it's not arguments anyway.

Jake-Moss committed:
Prefix DEBUG and ASAN env vars with "AEQ_", only enable undefined sanitiser on non-windows machines

pedrocamargo commented on aequilibrae/paths/cython/route_choice_set.pyx:
```diff
@@ -382,6 +383,140 @@ cdef class RouteChoiceSet:
+                "index"
+            ].reset_index()
+        except KeyError:
+            raise KeyError("not all origin and destinations IDs from the path files are present within the demand matrix")
```
All from the demand matrix MUST exist in the path files, but there is no need to exist demand for all OD pairs in the path file.  Think of the case where people prepare the choice sets for the ENTIRE zone system and then simply apply them to whatever matrix they have

Jake-Moss committed:
Flip the logic for checking demand values and ODs in the route set

Now if a value is non-zero within the demand matrix then it must be in the route set.

Jake-Moss committed:
Add warning for unreachable ODs in RC, change method of warning propagation

Jake-Moss committed:
Fix leaking the remaining items in the queue when we fill the route set
